src/Avalonia.X11/X11Window.cs

```diff
@@ -1133,6 +1133,46 @@ public PixelPoint Position
             }
         }
 
+        public IntPtr? ZOrder
```

Comment:
Enumerating all windows in the the entire desktop doesn't seem like a good idea.

Reply:
![image](https://github.com/AvaloniaUI/Avalonia/assets/1067584/ee8642b4-5696-4663-8959-1c75a7893fba)

Doesn't seem to be the case, windows are stacked as Konsole>Nemo>GEdit, while XQueryTree returns Nemo-GEdit-Konsole

Reply:
https://tronche.com/gui/x/xlib/window-information/XQueryTree.html

That's what the specification says:

> The children are listed in current stacking order, from bottommost (first) to topmost (last).

And that seem to be the case:

https://github.com/AvaloniaUI/Avalonia/assets/5689666/315c3a6c-6e51-4bca-956a-08eab22da72d

Those `nemo`, `gedit` and `Konsole` are something different than the windows themselves. Take a look that under `Qt Selection Owner for konsole` the order is correct and the windows have the correct title indicating those represent the actual windows. Maybe this is QT thing to comply with the specs? 


---


But I do agree maybe traversing the from root is not perfect. In that case, I guess in order to get correct z-order within the app, I have to traverse up the tree from each window, find a common parent, then find the index within the parent.

Reply:
@kekekeks Updated the implementation to traverse only those windows which are required to compute the relative (within the app) z-order.

Reply:
>And that seem to be the case

Might be WM-specific behavior. I'm running KWin 5.20.5 with XOrg 1.20.11, no wayland.